src/System.Private.CoreLib/src/System/Environment.CoreRT.Unix.cs

```diff
@@ -13,23 +14,70 @@ namespace System
     public static partial class Environment
     {
         internal static int CurrentNativeThreadId => ManagedThreadId.Current;
+
+        private static object _environmentLock = new object();
+        private static IDictionary<string, string> _environment = GetSystemEnvironmentVariables();
```

Comment:
This will be initialized as side-effect of calling `SetCommandLine` during startup. It tends to be surprising when you see it in the debugger. I am wondering whether it would make sense to have an explicit initialization of the value on the first actual call to one of the env functions.

Reply:
We can push it even further an initialize it on the first call to `SetEnvironmentVariableCore`, so that you pay for the copy of the environment only if you are setting the environment variables. Use the existing path otherwise.

Reply:
I knew that would happen, but it didn't seem like worth the effort. I'm happy to change it to lazy initialization tough. Were you thinking `LazyInitializer.EnsureInitialized` or some other trick?

Reply:
`Interlocked.CompareExchange`

Reply:
> Use the existing path otherwise.

That will inevitably lead to inconsistency w.r.t. P/Invoke calls to `setenv`. I deliberately wanted to avoid that based on https://github.com/dotnet/coreclr/issues/635 and the linked issues.

Reply:
It would find it weird if `setenv` was reflected at one point and not after a call to `SetEnvironmentVariable`. Since the initialization is lazy it can still happen, but at least in slightly less obscure way.

Reply:
> `setenv` was reflected at one point and not after a call to `SetEnvironmentVariable`

I would not feel bad about this. `setenv` in anything but a simple `C` program has unpredictable  behavior in number of ways.

Reply:
Ok, I will bring back the old code then... will have to bring back the `_environmentLock` too :-)